### DISCUSSION TITLE
What's the difference between `HighlightProvider.visualHighlightId` and `HighlightProvider.visualHighlightIds`?

### DISCUSSION THREAD

**Discussion by ollien** [vscode-neovim/vscode-neovim#2039] [Development]
I was taking another pass at cleaning up some of the highlighting code and I came across an old unresolved question of mine... 

In `HighlightProvider`, there are [two properties](https://github.com/vscode-neovim/vscode-neovim/blob/4890d3af03b5f93a46ffeadef1c4eed1fa0b2263/src/highlights/highlight_provider.ts#L118-L120), `visualHighlightId`, and `visualHighlightIds`. They seem to be [populated based on how many Visual highlight ids neovim sends us in a particular event](https://github.com/vscode-neovim/vscode-neovim/blob/4890d3af03b5f93a46ffeadef1c4eed1fa0b2263/src/highlights/highlight_provider.ts#L150-L153), but when used later, require a [highlight id to be set within both properties](https://github.com/vscode-neovim/vscode-neovim/blob/4890d3af03b5f93a46ffeadef1c4eed1fa0b2263/src/highlights/highlight_provider.ts#L210-L214) (which implies both properties are going to be set when setting up the highlight provider). 

I tried to narrow it down how they're used in practice a bit by looking at the values of `groups` and `id` ([here](https://github.com/vscode-neovim/vscode-neovim/blob/4890d3af03b5f93a46ffeadef1c4eed1fa0b2263/src/highlights/highlight_provider.ts#L150-L153)) as my editor launched, and I'm not sure if I understand the events coming in. Why are there three highlight ids all marked "Visual" (only one buffer is open), and yet only the last one will be stored as the `visualHighlightId` (given `groups` is of length 1 in each of these cases, no items will enter `visualHighlightIds`, but the property will be overwritten)? 

```
id=16 groups=['Visual']
id=108 groups=['Visual']
id=202 groups=['Visual']
```

I can understand wanting to perhaps remap all "Visual" highlight ids to one, but I don't think that's happening. Moreover, I'm not sure that I understand what these properties represent by themselves.  Can someone help me better understand what I'm looking at?

**Comment by xiyaowong** [comment]
Some syntax highlighting might be mixed with Visual, where the highlighting attributes are the same, but the highlighting IDs are different. This results in multiple decorations during calculations. When VSCode renders decorations, even if the attributes are identical, gaps may still appear in adjacent areas. Here, only Visual is handled because it is more commonly used. Also, see #1643

However, this issue has almost disappeared. First, we cleaned up Visual, and second, we linked all the highlights that needed to be cleared to VScodeNone, so there won't be many combined highlights. Nevertheless, it is worth keeping here.

> (given groups is of length 1 in each of these cases, no items will enter visualHighlightIds, but the property will be overwritten)

This is normal. We have many operations that adjust the highlight values, and the attributes are changing.

**Reply by ollien** [reply]
> Nevertheless, it is worth keeping here.

For any particular reason? I can't reason my way around a case that multiple items end up in this array, but if you know of one I'd love to see it. Is it just that syntax highlighting case?

> and the attributes are changing.

What attributes are changing?

**Reply by xiyaowong** [reply]
Try this patch

```diff
diff --git a/runtime/lua/vscode/highlight.lua b/runtime/lua/vscode/highlight.lua
index 7bd1dbc..dcba82f 100644
--- a/runtime/lua/vscode/highlight.lua
+++ b/runtime/lua/vscode/highlight.lua
@@ -24,7 +24,7 @@ api.nvim_set_hl(0, "VSCodeNone", {})
 ---@param value table
 local function set_hl(id, name, value)
   if vim.tbl_isempty(value) then
-    return api.nvim_set_hl(id, name, { link = "VSCodeNone" })
+    return api.nvim_set_hl(id, name, {})
   end
   return api.nvim_set_hl(id, name, value)
 end
@@ -52,7 +52,7 @@ local function setup_globals()
     Normal       = {},
     NormalNC     = {},
     NormalFloat  = {},
-    Visual       = {},
+    Visual       = { bg = '#abb2bf'},
     VisualNC     = {},
     VisualNOS    = {},
     Substitute   = {},
diff --git a/src/highlights/highlight_provider.ts b/src/highlights/highlight_provider.ts
index acd3e9c..b84dc3b 100644
--- a/src/highlights/highlight_provider.ts
+++ b/src/highlights/highlight_provider.ts
@@ -147,6 +147,7 @@ export class HighlightProvider implements Disposable {

     public addHighlightGroup(id: number, attrs: VimHighlightUIAttributes, groups: string[]): void {
         delete attrs.altfont;
+        console.log(id, attrs, groups);
         if (groups.includes("Visual")) {
             if (groups.length === 1) this.visualHighlightId = id;
             else this.visualHighlightIds.push(id);
```

**Reply by xiyaowong** [reply]
We need to reduce the number of decorators as much as possible because:
1. There are flaws in VSCode's rendering
2. They significantly affect performance.

**Reply by ollien** [reply]
Thanks, that was very helpful for my understanding. I guess this means we could remove it, so long as we aren't setting any color for `Visual`? Or am I missing your point

**Reply by xiyaowong** [reply]
You can't delete it just because it's not used now. This is correct logic, and there is no harm in keeping it

**Reply by ollien** [reply]
Yep, understood. It's sometimes a bit hard to refactor around dead code, I just wanted to make sure I was understanding right :)

Thanks!

**Comment by xiyaowong** [comment]
I've been adjusting the highlighting approach recently, which is likely to conflict with other changes. The remaining issues are not related to the current logic.

**Reply by ollien** [reply]
Cool, I'll hold off until I see that PR :) I have a few mental TODOs, mostly around code organization, but maybe your changes will help with that!
